src/Avalonia.Controls/Primitives/TextSelectionCanvas.cs

```diff
@@ -0,0 +1,361 @@
+        }
+
+        private Point ToTextBox(Point point)
+        {
+            return (_textBox is { } p) ? (p.VisualRoot as Visual)?.TranslatePoint(point, p) ?? point : point;
+        }
+
+        public void MoveHandlesToSelection()
+        {
+            if (_presenter == null || _textBox == null || _startHandle.IsDragging || _endHandle.IsDragging)
+            {
+                return;
+            }
+
+            var hasSelection = _textBox.SelectionStart != _textBox.SelectionEnd;
+            if (_caretHandle != null)
+            {
+                var points = _presenter.GetCaretPoints();
+
+                _caretHandle.SetTopLeft(ToLayer(points.Item2));
+            }
+
+            if (hasSelection)
+            {
+                var selectionStart = _textBox.SelectionStart;
+                var selectionEnd = _textBox.SelectionEnd;
+                var start = Math.Min(selectionStart, selectionEnd);
+                var length = Math.Max(selectionStart, selectionEnd) - start;
+
+                var rects = _presenter.TextLayout.HitTestTextRange(start, length);
+
+                if (rects.Any())
+                {
+                    var first = rects.First();
+                    var last = rects.Last();
+
+                    if (_startHandle != null && !_startHandle.IsDragging)
+                    {
+                        _startHandle.SetTopLeft(ToLayer(first.BottomLeft));
+                        _startHandle.SelectionHandleType = selectionStart < selectionEnd ? SelectionHandleType.Start : SelectionHandleType.End;
+                    }
+
+                    if (_endHandle != null && !_endHandle.IsDragging)
+                    {
+                        _endHandle.SetTopLeft(ToLayer(last.BottomRight));
+                        _endHandle.SelectionHandleType = selectionStart > selectionEnd ? SelectionHandleType.Start : SelectionHandleType.End;
+                    }
+                }
+            }
+        }
+
+        internal void SetPresenter(TextPresenter? textPresenter)
+        {
+            _presenter = textPresenter;
+            if (_presenter != null)
+            {
+                _textBox = _presenter.FindAncestorOfType<TextBox>();
+
+                if (_textBox != null)
+                {
+                    _textBox.AddHandler(TextBox.TextChangingEvent, TextChanged, handledEventsToo: true);
+                    _textBox.AddHandler(KeyDownEvent, TextBoxKeyDown, handledEventsToo: true);
+                    _textBox.AddHandler(LostFocusEvent, TextBoxLostFocus, handledEventsToo: true);
+                    _textBox.AddHandler(PointerReleasedEvent, TextBoxPointerReleased, handledEventsToo: true);
+                    _textBox.AddHandler(Gestures.HoldingEvent, TextBoxHolding, handledEventsToo: true);
+
+                    _textBox.PropertyChanged += _textBox_PropertyChanged;
+                    _textBox.LayoutUpdated += _textBox_LayoutUpdated;
```

Comment:
I know this was discussed during development, but `LayoutUpdated` is raised directly by the layout manager itself: even if some unrelated control in an unrelated part of the view has been re-arranged, the event will trigger for all textboxes, potentially causing performances problems.

It's needed here to know when the text box moves (e.g. it's inside a scroll viewer), to move the handles along with it.

I became aware two days ago of the `EffectiveViewportChanged` event. It should solve the problem nicely: it's only raised on the control itself, when its visible viewport has changed. It should be perfect for the use case here.

Reply:
EffectiveViewportChanged works. thanks.